pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryFullyQualifiedNameRule.java

```diff
@@ -256,12 +256,20 @@ private boolean isSamePackage(TypeNode node, String name) {
         }
 
         int i = name.lastIndexOf('.');
-        while (i > 0) {
+        if (i > 0) {
             name = name.substring(0, i);
             if (name.equals(currentPackage)) {
                 return true;
             }
-            i = name.lastIndexOf('.');
+        }
+        // if it is a name used inside a primary prefix, then it is ambiguous, whether it references
+        // a type or a field or a type inside a subpackage
+        // we assume here, it won't be a subpackage the name references a field, e.g.
+        // package a;
+        // name: a.b.c.d(); -> we assume, b is a class, c is a field, d is a method.
+        // but it could very well be, that: a.b is a package and c is a class, d is a (static) method.
+        if (node.jjtGetParent() instanceof ASTPrimaryPrefix) {
```

Comment:
There seem to be a couple of false-negatives now - PrimaryPrefix is maybe too restrictive...

* https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-web/src/test/java/org/springframework/protobuf/Msg.java#L80
* https://github.com/checkstyle/checkstyle/blob/checkstyle-8.10/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/design/finalclass/InputFinalClassClassWithPrivateCtorWithNestedExtendingClass.java#L12
* https://github.com/checkstyle/checkstyle/blob/checkstyle-8.10/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalType.java#L9